### PR TITLE
feat: add functionality to filter/display actions by user

### LINKED ISSUE #48: consolidate and display users across a space
light graasp users who use multiple subspaces or phases of a given graasp space generate a distinct user ID for each subspace they use; the goal is to introduce functionality to associate user IDs that belong to the same user. the expectation is that this will be done via the 'nickname' that a user provides when accessing a space. a user's consolidated activity across a space and its subspaces will then be displayed in the app dashboard

### PR DESCRIPTION
**Pull request by hagoptaminian**
using react-select; we can now see all users of a space, and display actions of selected users

fixes #48

### CODE REVIEW

File: src/components/custom/CustomValueContainer.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    const alsoSelectedCount = length - 2;
+    switch (length) {
+      case 1:
+        values = `${values[0].key}`;
+        break;
+      case 2:
+        values = `${values[0].key}, ${values[1].key}`;
+        break;
+      default:
+        values = `${values[0].key}, ${values[1].key}, and ${alsoSelectedCount} other${plural} selected`;
```
Comment (juancarlosfarah) [inline src/components/custom/CustomValueContainer.js:24]: Eventually we will have to incorporate translation here, but for now it's OK. Maybe add a `todo`

Comment (juancarlosfarah) [reply src/components/custom/CustomValueContainer.js:24]: And create an issue.

File: src/components/custom/CustomValueContainer.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import { components } from 'react-select';
+
+// this component, used within the UsersSelect component, overrides react-select's default text display
+// by default, react-select will keep expanding its text display as additional options are selected
+// with this component, after 2 selections, the text box will stop expanding and display: 'A, B, and X other(s) selected'
+// eslint-disable-next-line react/prop-types
+const CustomValueContainer = ({ children, ...props }) => {
```
Comment (juancarlosfarah) [inline src/components/custom/CustomValueContainer.js:8]: I really like this idea!

File: src/components/space/ActionsChart.js [hunk 3]
```diff
@@ -11,19 +11,65 @@ import {
...
+} from '../../utils/api';
 import { SpaceDataContext } from '../../contexts/SpaceDataProvider';
 
 const useStyles = makeStyles(() => ({
   typography: { textAlign: 'center' },
+  emptyChartAlert: {
+    display: 'flex',
+    justifyContent: 'center',
+    alignItems: 'center',
+    height: '450px',
```
Comment (juancarlosfarah) [inline src/components/space/ActionsChart.js:28]: Is this used elsewhere? Constant?

File: src/components/space/ActionsChart.js [hunk 4]
```diff
@@ -32,12 +78,12 @@ function ActionsChart() {
...
       <ResponsiveContainer width="95%" height={450}>
         <BarChart
-          data={actionsByDay}
+          data={formattedActionsByDay}
           margin={{ top: 30, bottom: 20, left: 20, right: 20 }}
         >
           <CartesianGrid strokeDasharray="2" />
-          <XAxis dataKey="date" />
-          <YAxis />
+          <XAxis dataKey="date" tick={{ fontSize: 14 }} />
```
Comment (juancarlosfarah) [inline src/components/space/ActionsChart.js:85]: Can we factor out these styling bits?

File: src/components/space/UsersSelect.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  );
+  const [selectedUsers, setSelectedUsers] = useState([]);
+
+  if (isLoading) {
+    return null;
+  }
+
+  // custom option allowing us to select all users in the dropdown
+  const allOption = {
+    name: 'Select all',
```
Comment (juancarlosfarah) [inline src/components/space/UsersSelect.js:57]: Translate and capitalise default as "Select All"

File: src/components/space/ActionsChart.js [hunk 6]
```diff
@@ -11,19 +11,65 @@ import {
...
   const { t } = useTranslation();
   const theme = useTheme();
   const classes = useStyles();
-  const { actions } = useContext(SpaceDataContext);
-  const actionsByDay = formatActions(getActionsByDay(actions));
+  const { actions, usersToFilter, allUsers } = useContext(SpaceDataContext);
+
+  // actionsByDay is the object passed, after formatting, to the BarChart component below
+  // if no users are selected (i.e. usersToFilter.length === 0), show all actions
+  // if all users are selected (i.e. usersToFilter.length === allUsers.length), also show all actions
```
Comment (juancarlosfarah) [inline src/components/space/ActionsChart.js:40]: Very nice comment. Makes everything clear.

File: src/components/space/UsersSelect.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    display: 'flex',
+    alignItems: 'center',
+  },
+  typography: {
+    marginRight: theme.spacing(1),
+  },
+}));
+
+// custom styling used by react-select (deviates from material-ui approach used in rest of app)
+const customStyles = {
```
Comment (juancarlosfarah) [inline src/components/space/UsersSelect.js:23]: Factor out some of the repeated values.

File: src/components/space/UsersSelect.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+  const handleChange = (selectedUser) => {
+    setSelectedUsers(selectedUser);
+    setUsersToFilter(selectedUser);
+  };
+
+  return (
+    <div className={classes.root}>
+      <Typography className={classes.typography}>
+        {t('Filter by user:')}
```
Comment (juancarlosfarah) [inline src/components/space/UsersSelect.js:69]: I prefer capitalising "Filter by User"

File: src/utils/api.js [hunk 9]
```diff
@@ -29,6 +29,34 @@ export const formatActions = (actionsByDayObject) => {
...
+};
+
+// given an actionsByDay object, findYAxisMax finds max value to set on the yAxis in the graph in ActionsChart.js
+export const findYAxisMax = (actionsByDay) => {
+  const arrayOfActionsCount = Object.values(actionsByDay);
+  if (!arrayOfActionsCount.length) {
+    return null;
+  }
+  const maxActionsCount = arrayOfActionsCount.reduce((a, b) => Math.max(a, b));
+  let yAxisMax;
```
Comment (juancarlosfarah) [inline src/utils/api.js:51]: Why do you have these two cases?

File: src/utils/api.js [hunk 10]
```diff
@@ -39,3 +67,63 @@ export const extractMainSpaceChildren = (arrayOfSpaceObjects) => {
...
+      consolidatedUsersArray[userIndex].ids.push(user._id);
+    }
+  });
+  return consolidatedUsersArray;
+};
+
+// for presentation purposes, format user names, then sort them alphabetically
+// proper names are recapitalized ('Augie March'), and only the first letter of emails is capitalized
+export const formatConsolidatedUsers = (consolidatedUsersArray) => {
+  const usersArrayCapitalized = consolidatedUsersArray.map((user) => {
```
Comment (juancarlosfarah) [inline src/utils/api.js:104]: Use `lodash` for this type of function.

File: src/utils/api.js [hunk 11]
```diff
@@ -39,3 +67,63 @@ export const extractMainSpaceChildren = (arrayOfSpaceObjects) => {
...
+      };
+    }
+    // otherwise, reg exp below capitalizes first letter of each part of a user's space-delimited name
+    return {
+      ...user,
+      name: user.name.replace(/\b(\w)/g, (char) => char.toUpperCase()),
+    };
+  });
+  // return alphbatically sorted array of users
+  return usersArrayCapitalized.sort((user1, user2) =>
```
Comment (juancarlosfarah) [inline src/utils/api.js:119]: Also use `lodash`

**Review by juancarlosfarah — CHANGES_REQUESTED** [review]
This is looking awesome! 🚀  

Just added some minor comments. Please have a look.

File: src/utils/api.js (hunk 9 above)
Comment (hagoptaminian) [reply src/utils/api.js:51]: if maxActionsCount <= 100 (e.g. we select a user who has 6 actions), we round up yAxisMax to the nearest ten, but if > 100, to the nearest hundred. I added a comment to clarify.

File: src/components/custom/CustomValueContainer.js (hunk 1 above)
Comment (hagoptaminian) [reply src/components/custom/CustomValueContainer.js:24]: I thought I caught everything needing an i18next wrapper with this PR but no 😅
I fixed this

File: src/components/space/ActionsChart.js (hunk 4 above)
Comment (hagoptaminian) [reply src/components/space/ActionsChart.js:85]: I experimented with this but Recharts does not respond well to such changes, so will leave as is (at least for now).

File: src/components/space/ActionsChart.js (hunk 3 above)
Comment (hagoptaminian) [reply src/components/space/ActionsChart.js:28]: Yes, this is also the height of the ResponsiveContainer which houses BarChart in this same component. Factored out the 450 to a `CONTAINER_HEIGHT` constant in `/config/constants`

**Comment by hagoptaminian** [comment]
@juancarlosfarah thanks for the detailed review 👍 pushed requested changes

**Review by juancarlosfarah — APPROVED** [review]

**Comment by juancarlosfarah** [comment]
@hagoptaminian, I approved this, BTW.

**Pull request merged**